pkg/apis/migration/v1alpha1/migtoken_types.go

```diff
@@ -0,0 +1,145 @@
+package v1alpha1
+
+import (
+	"context"
+	"errors"
+	"k8s.io/api/authentication/v1beta1"
+
+	authapi "k8s.io/api/authorization/v1"
+	kapi "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MigTokenSpec defines the desired state of MigToken
+type MigTokenSpec struct {
+	IdentitySecretRef *kapi.ObjectReference `json:"identitySecretRef"`
+	MigClusterRef     *kapi.ObjectReference `json:"migClusterRef"`
+}
+
+// MigTokenStatus defines the observed state of MigToken
+type MigTokenStatus struct {
+	Conditions
+	ObservedDigest string `json:"observedDigest,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MigToken is the Schema for the migtokens API
+// +k8s:openapi-gen=true
+type MigToken struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MigTokenSpec   `json:"spec,omitempty"`
+	Status MigTokenStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MigTokenList contains a list of MigToken
+type MigTokenList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []MigToken `json:"items"`
+}
+
+type Authorized map[string]bool
+
+func init() {
+	SchemeBuilder.Register(&MigToken{}, &MigTokenList{})
+}
+
+func (r *MigToken) HasRead(client k8sclient.Client, namespaces []string) (Authorized, error) {
+	authorized := Authorized{}
+	client, err := r.GetClient(client)
+	if err != nil {
+		return authorized, err
+	}
+	for _, namespace := range namespaces {
+		sar := authapi.SelfSubjectAccessReview{
+			Spec: authapi.SelfSubjectAccessReviewSpec{
+				ResourceAttributes: &authapi.ResourceAttributes{
+					Resource:  "namespaces",
+					Namespace: namespace,
+					Verb:      "get",
+				},
+			},
+		}
+		err := client.Create(context.TODO(), &sar)
+		if err != nil {
+			return authorized, err
+		}
+		authorized[namespace] = sar.Status.Allowed
+	}
+
+	return authorized, nil
+}
+
+func (r *MigToken) HasMigrate(client k8sclient.Client, namespaces []string) (Authorized, error) {
+	authorized := Authorized{}
+	_, err := r.GetClient(client)
+	if err != nil {
+		return authorized, err
+	}
+
+	return authorized, nil
+}
+
+func (r *MigToken) Authenticates(client k8sclient.Client) (bool, error) {
```

Comment:
Suggest `Authenticate()` since the method will perform (do) the authentication when called.